File: Solutions/606. Construct String from Binary Tree/606.py
class Solution(object):
    def tree2str(self, root):
      ans = []
      self.helper(root, ans)
      return ''.join(ans)
    
    def helper(self, root, ans):
      if root is None:
        return
      ans.append(str(root.val))
      if root.left is None and root.right is None:
        return
      ans.append('(')
      self.helper(root.left, ans)
      ans.append(')')
      if root.right is not None:
        ans.append('(')
        self.helper(root.right, ans)
        ans.append(')')
       

File: Solutions/606. Construct String from Binary Tree/606.java
class Solution {
    public String tree2str(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        helper(root, sb);
        return sb.toString();
    }
    public static void helper(TreeNode root, StringBuilder sb){
        if (root == null){
            return;
        }
        sb.append(String.valueOf(root.val));
        if (root.left == null && root.right == null) {
            return;
        }
        sb.append('(');
        helper(root.left, sb);
        sb.append(')');
        if(root.right != null) {
            sb.append('(');
            helper(root.right, sb);
            sb.append(')');
        }
    }
}